pulsar/src/main/java/com/hazelcast/jet/contrib/pulsar/PulsarSources.java

```diff
@@ -0,0 +1,137 @@
+     * @param projectionFn   converts a Pulsar message to an emitted item.
+     * @param <M>            the type of the message read by {@code PulsarConsumer}
+     * @param <T>            the type of data emitted from {@code StreamSource}
+     * @return a stream source to use in {@link Pipeline#readFrom}
+     */
+    public static <M, T> StreamSource<T> subscribe(
+            @Nonnull List<String> topics,
+            @Nonnull Map<String, Object> consumerConfig,
+            @Nonnull SupplierEx<PulsarClient> connectionSupplier,
+            @Nonnull SupplierEx<Schema<M>> schemaSupplier,
+            @Nonnull FunctionEx<Message<M>, T> projectionFn
+    ) {
+        checkSerializable(connectionSupplier, "connectionSupplier");
+        return SourceBuilder.timestampedStream("pulsar-ts-stream-source", ctx -> new PulsarSourceContext(
+                ctx.logger(), connectionSupplier.get(), topics, consumerConfig, schemaSupplier, projectionFn))
+                .<T>fillBufferFn(PulsarSourceContext::fillBuffer)
+                .destroyFn(PulsarSourceContext::destroy)
+                .build();
+    }
+
+    /**
+     * A context for the stream source of Apache Pulsar
+     *
+     * @param <M> the type of the value of message read by {@code PulsarConsumer}
+     * @param <T> the type of the emitted item after projection.
+     */
+    private static final class PulsarSourceContext<M, T> {
+        private final ILogger logger;
+        private final PulsarClient client;
+        private final Consumer<M> consumer;
+        private final FunctionEx<Message<M>, T> projectionFn;
+
+        private PulsarSourceContext(
+                @Nonnull ILogger logger,
+                @Nonnull PulsarClient client,
+                @Nonnull List<String> topics,
+                @Nonnull Map<String, Object> consumerConfig,
+                @Nonnull SupplierEx<Schema<M>> schemaSupplier,
+                @Nonnull FunctionEx<Message<M>, T> projectionFn
+        ) throws PulsarClientException {
+            checkSerializable(schemaSupplier, "schemaSupplier");
+            checkSerializable(projectionFn, "projectionFn");
+            this.logger = logger;
+            this.projectionFn = projectionFn;
+            this.client = client;
+            this.consumer = client.newConsumer(schemaSupplier.get())
+                                  .topics(topics)
+                                  .loadConf(consumerConfig)
+                                  .subscribe();
+        }
+
+        /**
+         * If there is an event time associated with the message, set the
+         * event time as the timestamp. Otherwise, it will set the publish time
+         * of this message as the timestamp.
+         */
+        private void fillBuffer(SourceBuilder.TimestampedSourceBuffer<T> sourceBuffer) {
+            CompletableFuture<Message<M>> messageFuture = consumer.receiveAsync();
+            messageFuture.thenAccept(message -> {
```

Comment:
We ignore if the future is completed exceptionally.

Reply:
By changing the receiving procedure, I removed the future logic.